cmd/vela-terraform/main.go

```diff
@@ -292,16 +294,24 @@ func run(c *cli.Context) error {
 		"registry": "https://hub.docker.com/r/target/vela-terraform",
 	}).Info("Vela Terraform Plugin")
 
-	// capture custom terraform version requested
-	version := c.String("terraform.version")
+	// capture custom terraform tfVersion requested
+	tfVersion := c.String("terraform.version")
 
-	// check if a custom terraform version was requested
-	if len(version) > 0 {
-		// attempt to install the custom terraform version
-		err := install(version, os.Getenv("PLUGIN_TERRAFORM_VERSION"))
+	// check if a custom terraform tfVersion was requested
+	if len(tfVersion) > 0 {
+		// attempt to install the custom terraform tfVersion
+		err := install(tfVersion, os.Getenv("PLUGIN_TERRAFORM_VERSION"))
 		if err != nil {
 			return err
 		}
+	} else {
+		// Default Version will be used if not defined
+		tfVersion = "1.0.0"
```

Comment:
If no version is specified, this will avoid the additional download and use the 1.0 behavior by default.

Reply:
Do we need this? We're already setting a "default" here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/Dockerfile#L8

That's the "default" version bundled with the image. If no custom version was defined, it uses that one.

---

FWIW, that value is also set here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/Dockerfile#L38

And then used here: https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/cmd/vela-terraform/main.go#L301 to avoid a download if someone specifies the same version the image was built with.

Reply:
> Do we need this? We're already setting a "default" here:

This is incorrect since the value of `c.String("terraform.version")` is only being populated via the environment variables of [PARAMETER_VERSION and TERRAFORM_VERSION](https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L148).

Reply:
hmm, maybe i'm missing something.

if `terraform.version` wasn't given it will skip the install and use the bundled "default" version that lives at https://github.com/go-vela/vela-terraform/blob/b74f9abd3a75ff152c3771d863b5b1f947b2463a/cmd/vela-terraform/terraform.go#L20 which is made available when the image is built.

if we're just talking about deriving the version value, can we just reference `PLUGIN_TERRAFORM_VERSION`'s value instead? otherwise, we are hard-coding default version values in a number of places now.

Reply:
> if we're just talking about deriving the version value, can we just reference PLUGIN_TERRAFORM_VERSION's value instead? otherwise, we are hard-coding default version values in a number of places now.

The current code changes requires that `terraform.version` is populated to allow the [SupportsChdir](https://github.com/go-vela/vela-terraform/blob/0bb13edc75df3c5664437e993839704773f8abd8/cmd/vela-terraform/main.go#L415) function to determine what logic to utilize. Do you think we should add `PLUGIN_TERRAFORM_VERSION` to the front of [this line](https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L148) to allow the user to override the default?

Reply:
IIRC: The ARG line in the Dockerfile is only available during the build.

Reply:
the ARG is, but we are populating the ENV with it in a later stage in the Dockerfile
```
✦ ❯ docker run -it --entrypoint sh target/vela-terraform
/ # env
HOSTNAME=42365df4871e
SHLVL=1
HOME=/root
PLUGIN_TERRAFORM_VERSION=1.0.0
TERM=xterm
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
/ # 
```

@JordanSussman sorry, for delay. if `terraform.version` isn't supplied, we could use `PLUGIN_TERRAFORM_VERSION` - wouldn't that work?

Reply:
> @JordanSussman sorry, for delay. if terraform.version isn't supplied, we could use PLUGIN_TERRAFORM_VERSION - wouldn't that work?

Are you saying that we should add `PLUGIN_TERRAFORM_VERSION` to the list of parsed environment variables that urfavecli utilizes to set the value of `terraform.version`?

https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L147-L152

I would be happy to append `PLUGIN_TERRAFORM_VERSION` to the end of the existing list.

Reply:
As long as the user supplied value overrides it, i don't see a downside?

Reply:
Check https://github.com/JayCeeJr/vela-terraform/commit/d72ed96ed8617e06f68d6001b4c6ff83e6985cce
I think this will work and use the default version if not specified

Reply:
@JayCeeJr you should be able to remove https://github.com/JayCeeJr/vela-terraform/commit/d72ed96ed8617e06f68d6001b4c6ff83e6985cce in favor of adding `PLUGIN_TERRAFORM_VERSION` to the beginning of https://github.com/go-vela/vela-terraform/blob/c9d58f8110825cccefc1fad0a395cad8e1a6a397/cmd/vela-terraform/main.go#L148.

Reply:
@JordanSussman , I'm fairly certain we don't want to do that.

If we do that won't `c.String("terraform.version")  = os.Getenv("PLUGIN_TERRAFORM_VERSION")`
https://github.com/go-vela/vela-terraform/blob/d72ed96ed8617e06f68d6001b4c6ff83e6985cce/cmd/vela-terraform/main.go#L298-L304

Which will make this always true 
https://github.com/go-vela/vela-terraform/blob/8930ba57f593c8731699d231198bf12fe62658e1/cmd/vela-terraform/terraform.go#L32-L36
and never install a new version.

That being said, maybe just setting the default for the arg to match the preinstalled version will skip the install anyway.

Reply:
Are the EnvVars applied in order if multiples are specified?... I guess that would be the same as what I was suggesting.

Reply:
> Are the EnvVars applied in order if multiples are specified?... I guess that would be the same as what I was suggesting.

I wrote a brief script to validate the behavior of urfavecli:

```golang
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/urfave/cli/v2"
)

func main() {
	app := &cli.App{
		Name:  "boom",
		Usage: "make an explosive entrance",
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:    "test",
				EnvVars: []string{"TEST_VAR_1", "TEST_VAR_2", "TEST_VAR_3"},
			},
		},
		Action: func(c *cli.Context) error {
			fmt.Println(c.String("test"))
			return nil
		},
	}

	err := app.Run(os.Args)
	if err != nil {
		log.Fatal(err)
	}
}
```

```sh
TEST_VAR_2=test2 TEST_VAR_3=test3 TEST_VAR_1=test1 go run main.go
```

TL;DR urfave will utilize the first environment variable within the EnvVars slice that is populated, so you should be safe to append `PLUGIN_TERRAFORM_VERSION` to the end of the existing slice.

Reply:
Added to EnvVars
https://github.com/go-vela/vela-terraform/blob/6d75345867705b86b776604e6fc4a22e523efd96/cmd/vela-terraform/main.go#L149-L154